### ISSUE TITLE
[pt20][aot_eager] Exceed Python recursion limit with huge model or frequent recompilation

### ISSUE THREAD

**Issue by C1rN09** [pytorch/pytorch#93935]
### 🐛 Describe the bug

`aot_eager` backend wraps `bw_compiler` constantly on each graph break & recompilation, making the function call exceeds Python recursion limit for huge model or frequent recompilations. Repro example below

```python
import torch
import torch.nn as nn
import torch._dynamo as dynamo
from torch._dynamo import disallow_in_graph
import random

# manually simulate graph breaks
def graph_break():
    pass
disallow_in_graph(graph_break)

class Repro(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear1 = nn.Linear(4, 4)
        self.linear2 = nn.Linear(4, 4)
        self.linear3 = nn.Linear(4, 4)
    def forward(self, x):
        x = self.linear1(x)
        graph_break()
        x = self.linear2(x)
        graph_break()
        x = self.linear3(x)
        return x

if __name__ == '__main__':
    model = Repro().cuda()
    opt = torch.compile(model, backend='aot_eager')
    # variable batch_size to simulate recompilations
    for batch_size in range(3):
        data = torch.randn(batch_size, 4).cuda()
        out = opt(data)
        loss = out.sum()
        loss.backward()
```

Add the following lines to [torch/_dynamo/optimizations/training.py](https://github.com/pytorch/pytorch/blob/a2fded30012e26d8c469d2b668a226315794a559/torch/_dynamo/optimizations/training.py#L54-L56) and then run the above script, the results as follow

```python
def _wrapped_bw_compiler(*args, **kwargs):
    global _count
    _count += 1
    print(f'{_count}: {bw_compiler}')
    return eval_frame.disable(eval_frame.disable(bw_compiler)(*args, **kwargs))
```
![image](https://user-images.githubusercontent.com/112053249/216346230-bce5816c-fc15-47b4-990d-f15ff71ef3fe.png)

Is this intended or a bug? What else can I do apart from reducing `torch._dynamo.config.cache_size_limit` and increasing `sys.setrecursionlimit()` to prevent this issue?

### Versions

```text
PyTorch version: 2.0.0.dev20230201+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 8.5.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.17

Python version: 3.8.15 (default, Nov 24 2022, 15:19:38)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-3.10.0-957.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.6.124
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB
GPU 3: NVIDIA A100-SXM4-80GB
GPU 4: NVIDIA A100-SXM4-80GB
GPU 5: NVIDIA A100-SXM4-80GB
GPU 6: NVIDIA A100-SXM4-80GB
GPU 7: NVIDIA A100-SXM4-80GB

Nvidia driver version: 515.65.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] pytorch-triton==2.0.0+0d7e753227
[pip3] torch==2.0.0.dev20230201+cu116
[pip3] torchaudio==2.0.0.dev20230112+cu116
[pip3] torchvision==0.15.0.dev20230131+cu116
[conda] numpy                     1.24.1                   pypi_0    pypi
[conda] pytorch-triton            2.0.0+0d7e753227          pypi_0    pypi
[conda] torch                     2.0.0.dev20230201+cu116          pypi_0    pypi
[conda] torchaudio                2.0.0.dev20230112+cu116          pypi_0    pypi
[conda] torchvision               0.15.0.dev20230131+cu116          pypi_0    pypi
```

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @mlazos @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

**Comment by C1rN09** [comment]
Believe this is a bug in `aot_eager`, because the following code **failed `aot_eager` but passed `inductor`**
```python
import torch
import torch.nn as nn
import torch._dynamo.config
from torch._dynamo import disallow_in_graph
import sys

# 6 graph breaks x 50 recompilations, exceeds 300 recursion limit
sys.setrecursionlimit(300)

# manually simulate graph breaks
def graph_break():
    pass
disallow_in_graph(graph_break)

# A model with 6 graph breaks
class Repro(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear1 = nn.Linear(4, 4)
        self.linear2 = nn.Linear(4, 4)
        self.linear3 = nn.Linear(4, 4)
        self.linear4 = nn.Linear(4, 4)
        self.linear5 = nn.Linear(4, 4)
        self.linear6 = nn.Linear(4, 4)
    def forward(self, x):
        x = self.linear1(x)
        graph_break()
        x = self.linear2(x)
        graph_break()
        x = self.linear3(x)
        graph_break()
        x = self.linear4(x)
        graph_break()
        x = self.linear5(x)
        graph_break()
        x = self.linear6(x)
        return x

if __name__ == '__main__':
    model = Repro().cuda()
    opt = torch.compile(model, backend='aot_eager')
    # opt = torch.compile(model, backend='inductor')  # inductor gives no error
    # variable batch_size to simulate recompilations
    for batch_size in range(50):
        data = torch.randn(batch_size, 4).cuda()
        out = opt(data)
        loss = out.sum()
        loss.backward()
```

I know that `aot_eager` is just for debugging, but this bug may be misleading for debugging, since it may throw different errors for `inductor` and `aot_eager` mode.

**Comment by jon-chuang** [comment]
@C1rN09 , Some observations:
1. I can reproduce with just `range(25)`, so I think the recursion happens twice within a single graph break. (or perhaps once each for the forward and backward subgraphs)
2. This does not trigger if I do not call `loss.backward()`. As a matter of fact, the recursion occurs within `loss.backwards()`. What's going on here? The answer is that when we run `autograd.backward()`, it actually calls our `compiled_fn`s, which, as we recall, are actually wrapped into callbacks in which we recursive create `DisableContext` wrappers. When we graph break due to a guard failure, I believe we recompile our already compiled code. I haven't fully investigated how it is that we manage to nest the successively compiled functions, or what role backwards has to play in this. I will update as I investigate.

Some things I'm wondering:
1. (Question for PyTorch team) How come AOT autograd doesn't support dynamic shapes, resulting in constant recompilation? We already have infrastructure for dynamic shapes via `SymInt`.

Overall, I agree this is a bug, we should not need to have so many recursive calls on recompilation.

**Comment by ezyang** [comment]
we actually have it working here https://github.com/pytorch/pytorch/pull/93059 but you need to keep bugging @Chillee to land it lol

**Comment by ezyang** [comment]
Training support for dynamic shapes is landed in nightlies, give it a try

**Comment by williamwen42** [comment]
Did this get fixed? I find that `_count` in the repro doesn't grow after a certain number of iterations with dynamic shapes enabled (also need to increase cache size limit to see difference between dynamic shapes on/off).
